simple_certmanager/utils.py

```diff
@@ -5,20 +5,65 @@
 from django.utils.encoding import force_str
 
 from cryptography import x509
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric.types import PrivateKeyTypes
 from cryptography.hazmat.primitives.serialization import load_pem_private_key
 
 logger = logging.getLogger(__name__)
 
 
-def load_pem_x509_private_key(data: bytes):
+class PrivateKeyError(Exception):
+    pass
+
+
+class KeyIsEncrypted(PrivateKeyError):
+    pass
+
+
+class BadPassword(PrivateKeyError):
+    pass
+
+
+def load_pem_x509_private_key(data: bytes, password: str | None = None):
     """
     Small wrapper around the ``cryptography.hazmat`` private key loader.
 
     Nothing in this code is really specific to x509, but the function name expresses
     our *intent* to only deal with x509 certificate/key pairs.
     """
-    # passwords are not supported
-    return load_pem_private_key(data, password=None)
+    _password = password.encode("utf-8") if password is not None else None
+    try:
+        return load_pem_private_key(data, password=_password)
+    except TypeError as exc:
+        if (
+            password is None
+            and exc.args
+            and (err := exc.args[0])
+            and "Password was not given but private key is encrypted" in err
+        ):
+            raise KeyIsEncrypted("Private key is encrypted") from exc
+        # an error we can't handle, so just re-raise and let it bubble up
+        else:
+            raise
+    except ValueError as exc:
+        if (
+            password is not None
+            and exc.args
+            and (err := exc.args[0])
+            and "Incorrect password?" in err
+        ):
+            raise BadPassword("Could not decrypt with the given password") from exc
+        # an error we can't handle, so just re-raise and let it bubble up
```

Comment:
Looking at https://github.com/pyca/cryptography/issues/8563 we may have to catch more error types

Reply:
https://cryptography.io/en/latest/hazmat/primitives/asymmetric/serialization/#cryptography.hazmat.primitives.serialization.load_pem_private_key probably just need to catch `ValueError` and `TypeError` and treat them without looking at the error messages themselves.

Reply:
self-reminder: pyca/cryptography#11291 confirms that we should only target `TypeError` and `ValueError`, the error strings are not guaranteed to stay the same.